### ISSUE TITLE
NullReferenceException on TaskAwaiter`1.get_IsCompleted() when using with grpc-dotnet client

### ISSUE THREAD

**Issue by Mitch528** [microsoft/coyote#180]
Exception output:

```
System.NullReferenceException
Object reference not set to an instance of an object.
   at Microsoft.Coyote.Tasks.TaskAwaiter`1.get_IsCompleted()
   at CoyoteRepro.Tests.CoyoteUnitTest.Coyote_Should_Not_Throw_NRE() in /Users/mitch528/repos/CoyoteRepro/CoyoteRepro/CoyoteRepro.Tests/CoyoteUnitTest.cs:line 42
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_1.<<InvokeTestMethodAsync>b__1>d.MoveNext() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestInvoker.cs:line 264
--- End of stack trace from previous location ---
   at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\ExecutionTimer.cs:line 48
   at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in C:\Dev\xunit\xunit\src\xunit.core\Sdk\ExceptionAggregator.cs:line 90
```

Repro: https://github.com/Mitch528/CoyoteRepro

Expected:

Both tests should succeed

Actual:

Test awaiting `Task<TResponse>` succeeds, but test awaiting `AsyncUnaryCall<TResponse>` fails ([code](https://github.com/Mitch528/CoyoteRepro/blob/main/CoyoteRepro/CoyoteRepro.Tests/CoyoteUnitTest.cs#L33)).

**Comment by pdeligia** [comment]
Thanks for reporting this issue @Mitch528 and extra thanks for providing a repro!

To help us debug this, could you let us know how you invoke the concurrent tests with Coyote? So I see that you are using xUnit as your unit test runner, which is great, but I can't see any use of the Coyote `TestingEngine` which is required to actually run the tests with Coyote after they have been rewritten and also configures how these tests should run (e.g. for N iterations). Unless I am missing something in your repro repo?

How to use the `TestingEngine` API is explained in this doc: https://microsoft.github.io/coyote/#how-to/unit-testing/.

Now, if the tests run _after_ the assembly has been rewritten but _without_ the `TestingEngine`, then they will run as regular (non-Coyote) tests but the Coyote hooks are already inserted. We normally expect people to run rewritten tests the engine, but these hooks should be pretty much no-op or pass-through without Coyote, so its interesting that you are getting this NRE and should get to the bottom of it. But letting you know about the `TestingEngine` in case you have not seen this requirement.

**Comment by Mitch528** [comment]
@pdeligia No problem. Oh, I didn't realize that. I've updated the repro to use TestingEngine, but I'm still getting the same exception.

```
..... Iteration #1
[xUnit.net 00:00:02.79]     CoyoteRepro.Tests.CoyoteUnitTest.Coyote_Should_Not_Throw_NRE [FAIL]
  Failed CoyoteRepro.Tests.CoyoteUnitTest.Coyote_Should_Not_Throw_NRE [139 ms]
  Error Message:
   Coyote found 1 bug(s) in Coyote_Should_Not_Throw_NRE.
Expected: True
Actual:   False
  Stack Trace:
     at CoyoteRepro.Tests.CoyoteUnitTest.RunCoyoteTest(ITestOutputHelper output, Func`1 test, String testName) in /Users/mitch528/repos/CoyoteRepro/CoyoteRepro/CoyoteRepro.Tests/CoyoteUnitTest.cs:line 95
   at CoyoteRepro.Tests.CoyoteUnitTest.Coyote_Should_Not_Throw_NRE() in /Users/mitch528/repos/CoyoteRepro/CoyoteRepro/CoyoteRepro.Tests/CoyoteUnitTest.cs:line 43
  Standard Output Messages:
 Coyote found 1 bug(s).
 Bug Reports: Object reference not set to an instance of an object.



Failed!  - Failed:     1, Passed:     1, Skipped:     0, Total:     2, Duration: 1 s - /Users/mitch528/repos/CoyoteRepro/CoyoteRepro/CoyoteRepro.Tests/bin/net5.0/CoyoteRepro.Tests.dll (net5.0)
```

**Comment by pdeligia** [comment]
No worries, and yes, I am able to also repro it from my side too with or without the `TestingEngine`. It seems like a weird corner case, and I have a suspicion why it might be happening, and trying to figure it out. Thanks for the repro, this was incredibly useful!

Btw, as a heads up: you will need to use the `WithNoBugTraceRepro()` option in your test `Microsoft.Configuration`, for example:
```csharp
var config =Configuration.Create().WithTestingIterations(100).WithNoBugTraceRepro();
var engine = TestingEngine.Create(config, test.Coyote_Should_Succeed);
engine.Run();
```

This is explained in this [doc page](https://microsoft.github.io/coyote/#get-started/using-coyote#supported-scenarios). Basically, because your test goes through ASP.NET assemblies which are not (or I should say cannot be) rewritten (because they contain currently unsupported concurrency APIs), the tool has to give up repro to be able to test such "partially" controlled concurrency. You still get testing and can find bugs though, just that you cannot replay their traces using the Coyote bug replay mode.

If you don't enable this option, and try to run say `Coyote_Should_Succeed`, you will get this error + explanation:
```
Method 'Grpc.Core.AsyncUnaryCall.get_ResponseAsync' returned an uncontrolled task with id '3', which is not allowed by default as it can interfere with the ability to reproduce bug traces: either mock the method returning the uncontrolled task, or rewrite its assembly. Alternatively, use the '--no-repro' command line option to ignore this error by disabling bug trace repro. Learn more at http://aka.ms/coyote-no-repro.
```

Hope this helps! In the meantime, I will let you know once I have managed to figure out the above error and hopefully can include a fix in a new release ASAP.

**Comment by pdeligia** [comment]
@Mitch528 just to update you, I have a fix for this issue in #181, will try to merge and release a new package soon.

**Comment by pdeligia** [comment]
This has been fixed in [v1.2.8](https://www.nuget.org/packages/Microsoft.Coyote/1.2.8) so closed the issue, but let us know if you face more issues.
